### PR TITLE
Adding Selecting Chain Interaction And Flow Improvements

### PR DESCRIPTION
**Pull request by rabi-siddique**
The PR does the following:
- Add interaction in the `importWallet` flow to select chain. 
- Introduces some helper methods to add reliable test cases. 
- Address test cases giving different outputs based on connection speed. 
- Changes in the flow to get wallet address.

### CODE REVIEW

File: tests/e2e/specs/keplr/keplr-spec.js [hunk 1]
```diff
@@ -87,10 +97,13 @@ describe('Keplr', () => {
...
       cy.getTokenAmount('ATOM').then(tokenValue => {
-        expect(tokenValue).to.equal(0);
+        expect(Number.isFinite(tokenValue)).to.be.true;
       });
       cy.getTokenAmount('BLD').then(tokenValue => {
-        expect(tokenValue).to.equal(331);
+        expect(Number.isFinite(tokenValue)).to.be.true;
+      });
+      cy.getTokenAmount('IST').then(tokenValue => {
+        expect(Number.isFinite(tokenValue)).to.be.true;
```
Comment (frazarshad) [inline tests/e2e/specs/keplr/keplr-spec.js:106]: revert this so that our tests are more robust

File: tests/e2e/specs/keplr/keplr-spec.js [hunk 2]
```diff
@@ -99,10 +112,14 @@ describe('Keplr', () => {
       });
     });
 
-    it(`should get wallet address`, () => {
-      cy.getWalletAddress().then(walletAddress => {
+    it(`should get wallet address without running addNewTokensFound flow`, () => {
+      cy.getWalletAddress('AGORIC').then(walletAddress => {
+        expect(walletAddress.length).to.be.equal(45);
+      });
```
Comment (frazarshad) [inline tests/e2e/specs/keplr/keplr-spec.js:118]: exact match address

File: pages/keplr/notification-page.js [hunk 3]
```diff
@@ -1,9 +1,15 @@
 const approveButton = `button`;
 const copyAddress = 'Copy Address';
-const copyWalletAddressSelector = 'div.sc-dkzDqf div.sc-hKMtZM.sc-kDDrLX.cyoEAq.dkJSBQ'
+const copyWalletAddressSelector =
+  'div.sc-dkzDqf div.sc-hKMtZM.sc-kDDrLX.cyoEAq.dkJSBQ';
```
Comment (frazarshad) [inline pages/keplr/notification-page.js:4]: lets remove this if unused

File: commands/keplr.js [hunk 4]
```diff
@@ -238,12 +267,20 @@ const keplr = {
...
+    playwright.switchToKeplrWindow();
     await module.exports.goToHome();
-    const page = playwright.keplrWindow();
+    const newTokensSelctorExists = await playwright.waitForAndCheckElementExistence(
+      homePageElements.newTokensFoundSelector,
+    );
+
+    if (newTokensSelctorExists) {
+      await module.exports.addNewTokensFound(false);
+    }
```
Comment (frazarshad) [inline commands/keplr.js:279]: dont think we need to do this internally. in most cases this will not be needed
in cases when it is needed, a user will know that tokens are not currently present and can call it themselves explicitly

File: pages/keplr/notification-page.js [hunk 5]
```diff
@@ -1,9 +1,15 @@
...
+const walletSelectors = {
+  AGORIC: 'img[alt="Agoric local"]',
+  COSMOS: 'img[alt="Cosmos Hub"]',
+};
```
Comment (frazarshad) [inline pages/keplr/notification-page.js:8]: we should have a function like so
`const walletSelector = (chainName) => 'img[alt="${chainName}]'`
so that we can dynamically generate a selector, instead of manually adding a selector for every chain

File: pages/keplr/first-time-flow-page.js [hunk 6]
```diff
@@ -14,12 +14,17 @@ const phraseAccountCreated = 'Account Created!';
 const finishButton = 'button[type="button"]';
 const textAreaSelector = 'textbox';
 const submitPhraseButton = 'button[type="submit"]';
+const chainName = 'Agoric local'
+const chainNameSelector = 'text=Agoric local'
```
Comment (frazarshad) [inline pages/keplr/first-time-flow-page.js:18]: we shouldnt have selectors this specific

File: commands/keplr.js [hunk 7]
```diff
@@ -170,6 +167,38 @@ const keplr = {
...
+          playwright.keplrWindow(),
+        );
+      }
+    } else {
+      await playwright.waitAndClick(
+        onboardingElements.submitChainButton,
+        playwright.keplrWindow(),
+      );
+    }
+  },
```
Comment (frazarshad) [inline commands/keplr.js:201]: i think we can adapt this to handle chains dynamically i.e a user could provide us with an array ['Agoric', 'Cosmos hub', 'Secret'] and we could select all chains that match that name

File: commands/keplr.js [hunk 8]
```diff
@@ -287,22 +324,25 @@ const keplr = {
...
     await playwright.waitAndClickByText(
       homePageElements.addChainsButton,
       playwright.keplrWindow(),
       true,
     );
-    await playwright.switchToCypressWindow();
+
+    if (switchScreens) {
+      await playwright.switchToCypressWindow();
+    }
```
Comment (frazarshad) [inline commands/keplr.js:345]: i think we should always be switching screens regardless. 
currently we're building our flows with the mindset that every interaction/command will switch to the keplr window, perform its interaction, and switch back to the cypress window.
if we start performing interactions that deviate from that, it could break the flow of subsequent interactions

File: commands/keplr.js [hunk 9]
```diff
@@ -319,10 +359,22 @@ const keplr = {
...
+        break;
+      }
+    }
 
+    if (!numberString) {
+      throw new Error(`Token value for ${tokenName} not found`);
+    }
+
+    let tokenValue = parseFloat(numberString);
+    await playwright.switchToCypressWindow();
```
Comment (frazarshad) [inline commands/keplr.js:377]: i dont think we need to make it this complicated:
```
const tokenValue = textArray[3];
const parsedTokenValue = Number(element.replace(/,/g, ''));
await playwright.switchToCypressWindow();
```
should suffice

File: commands/playwright-keplr.js [hunk 10]
```diff
@@ -307,6 +308,29 @@ module.exports = {
...
+    while (retries < maxRetries) {
+      try {
+        await module.exports.waitAndClick(
+          selector,
+          module.exports.keplrWindow(),
+          options,
+        );
+        return;
+      } catch (error) {
+        retries++;
```
Comment (frazarshad) [inline commands/playwright-keplr.js:324]: we should add a small delay between retries to avoid differences between slow and fast computes
```
page.waitForTimeout(100);
```

File: commands/playwright-keplr.js [hunk 11]
```diff
@@ -307,6 +308,29 @@ module.exports = {
...
+      }
+    }
+
+    throw new Error(`Failed to click element after ${maxRetries} attempts`);
+  },
+  async waitAndClickWithDelay(selector, options, delay) {
+    await new Promise(resolve => setTimeout(resolve, delay));
```
Comment (frazarshad) [inline commands/playwright-keplr.js:331]: we can use `page.waitForTimeout()` instead. its better to use the function provided by playwright for this + we use it in other places as well so it would be consistent

Comment (rabi-siddique) [reply commands/playwright-keplr.js]: Nice catch. Done

File: commands/playwright-keplr.js (hunk 10 above)
Comment (rabi-siddique) [reply commands/playwright-keplr.js:324]: Introducing retries was done to cater slow and fast computes. Adding delay will slow things down

File: commands/keplr.js (hunk 8 above)
Comment (rabi-siddique) [reply commands/keplr.js:345]: I needed to reuse this implementation in a flow I was implementing without the screen-switching part. All the tests were passed and now flow was broken.

File: commands/keplr.js (hunk 7 above)
Comment (rabi-siddique) [reply commands/keplr.js:201]: Agreed. We can add that in a separate PR or add to our TODOs.

File: pages/keplr/notification-page.js (hunk 5 above)
Comment (rabi-siddique) [reply pages/keplr/notification-page.js]: Great suggestion. Done.

File: pages/keplr/first-time-flow-page.js (hunk 6 above)
Comment (rabi-siddique) [reply pages/keplr/first-time-flow-page.js:18]: They're used for selecting chain and not giving any issues.

File: commands/keplr.js (hunk 4 above)
Comment (rabi-siddique) [reply commands/keplr.js:279]: I don't think that is reliable. We should not assume that the user knows. A flow should be capable of handling all edge cases.

File: tests/e2e/specs/keplr/keplr-spec.js (hunk 1 above)
Comment (rabi-siddique) [reply tests/e2e/specs/keplr/keplr-spec.js:106]: It shows different coins on our computers plus what if the contract gets updated and the coin values change? I feel in the tests for workflows, hard coding the value will not help.

Comment (frazarshad) [reply tests/e2e/specs/keplr/keplr-spec.js:106]: the tests are designed for the CI not for a personal device. on the CI the value will be exactly the same everytime.
Our tests are (and should be) deterministic i.e the values at a any point of the test should be known exactly.
If the values change then that is a failure on our part and we need to fix the tests or the setup acordingly

File: commands/keplr.js (hunk 4 above)
Comment (frazarshad) [reply commands/keplr.js:279]: this is not an edge case, the writer of a test should know if the wallet address is present or not before writing the test.

There can be cases where user does not want to press the 'add new token' button until a later test. and since our code is forcing him to add new tokens, it might ruin their flow.

Plus all our interactions should have a single fuctionality/responsibility. here we are performing two different function together which a user can always perform separately

File: pages/keplr/first-time-flow-page.js (hunk 6 above)
Comment (frazarshad) [reply pages/keplr/first-time-flow-page.js:18]: but not every user wants the agoric chain in their setup

File: commands/keplr.js (hunk 8 above)
Comment (frazarshad) [reply commands/keplr.js:345]: what flow was this? i dont see it being used anywhere
the `should get the accurate values for the tokens in the wallet` spec uses `getTokenAmount` twice in succession and it doesn't have any issues.

File: commands/playwright-keplr.js (hunk 10 above)
Comment (frazarshad) [reply commands/playwright-keplr.js:324]: retries will run very quickly on a fast computer, but very slowly on a slow computer.
lets say an element appears after 300ms 
the slow computer might perform 5 retires in 350ms and find the item at the final retry
the fast computer might run the retries in 50ms and miss all the retires

File: tests/e2e/specs/keplr/keplr-spec.js (hunk 1 above)
Comment (rabi-siddique) [reply tests/e2e/specs/keplr/keplr-spec.js:106]: I agree. Tests should be deterministic. `expect(Number.isFinite(tokenValue)).to.be.true;` is not only deterministic but it also future-proofs the test suite by making it more adaptable to changes in requirements(contract).

File: commands/keplr.js (hunk 4 above)
Comment (rabi-siddique) [reply commands/keplr.js:279]: This is an edge case. 

Scenario:
The user want's agoric wallet address. Goes to copy the address but the agoric chain isn't present there ==> Mainly because the user did not include new tokens ==> Output is the user never got the address when they should've gotten it.

Also, assumptions lead to incorrect behaviour and unreliable software.  

As for SRP, I am all in for it. But then we should apply it consistently and where it's required.

File: commands/keplr.js (hunk 8 above)
Comment (rabi-siddique) [reply commands/keplr.js:345]: Please see the `getWalletAddress` function

File: commands/keplr.js (hunk 9 above)
Comment (rabi-siddique) [reply commands/keplr.js:377]: Based on the string we are grabbing and extracting the number, if we analyze it, it just contains one number. 
This is working and giving accurate results. For the new UI, we are not sure what it will contain or what changes Keplr will introduce. We don't control the UI Keplr and in that sense, both approaches are bad then.

File: tests/e2e/specs/keplr/keplr-spec.js (hunk 1 above)
Comment (rabi-siddique) [reply tests/e2e/specs/keplr/keplr-spec.js:106]: @frazarshad Thank You for taking the initiative and having the meeting with me to explain the issue going with this approach. Fixed.

File: tests/e2e/specs/keplr/keplr-spec.js [hunk 12]
```diff
@@ -75,6 +75,16 @@ describe('Keplr', () => {
         expect(setupFinished).to.be.true;
       });
     });
+
+    it(`should get wallet address while running addNewTokensFound flow`, () => {
+      cy.getWalletAddress('AGORIC').then(walletAddress => {
+        expect(walletAddress.length).to.be.equal(45);
+      });
+
```
[the remark this answers is not in the document]
Comment (frazarshad) [reply tests/e2e/specs/keplr/keplr-spec.js]: nvm. chain address actually changes between runs

File: tests/e2e/specs/keplr/keplr-spec.js [hunk 13]
```diff
@@ -75,6 +75,16 @@ describe('Keplr', () => {
...
+    it(`should get wallet address while running addNewTokensFound flow`, () => {
+      cy.getWalletAddress('Agoric localhost').then(walletAddress => {
+        expect(walletAddress.length).to.be.equal(45);
+      });
+
+      cy.getWalletAddress('Cosmos Hub').then(walletAddress => {
+        expect(walletAddress.length).to.be.equal(45);
+      });
+    });
```
Comment (frazarshad) [inline tests/e2e/specs/keplr/keplr-spec.js:87]: @rabi-siddique  we have two sets of these test cases. both which don't technically run the addNewTokensFound flow. we should fix this in a later PR

**Review by frazarshad — APPROVED** [review]

File: tests/e2e/specs/keplr/keplr-spec.js (hunk 13 above)
Comment (rabi-siddique) [reply tests/e2e/specs/keplr/keplr-spec.js:87]: @frazarshad The test cases you're referring to are working fine. `addNewTokensFound` flow is essential for reliable results.
